Support `~/.aws/config` file parsing

AliSoftware opened the pull request:
## What

Especially so that even if `region` is not defined in `~/.aws/credentials` but is defined in `~/.aws/config`, we can still create `AWSCredentials` with that `region` value.

Fixes https://github.com/Automattic/hostmgr/issues/83

## How

The parsing which was previously done by the `AWSCredentialsFileParser` class is now handled via:
 - A `AWSProfileConfig` class, used to parse *any* AWS user config file—be it `~/.aws/credentials` as before, or `~/.aws/config` (new)—and collect all the key/value pairs of each profile found in those files.
 - The `AWSCredentials` class can then now be constructed from an array of those `AWSProfileConfig` objects, combining the values for a profile parsed from different files if needed

I've then updated the tests to account for more scenarios combining those.

> [!Note]  
> While the `~/.aws/credentials` and `~/.aws/config` files are very similar, they use [slightly different format for profile sections](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-format), with the `~/.aws/config` file listing profiles as `[profile <profilename>]`—except for the special `[default]` one—and potentially having other section types, like `[services <…>]` or `[sso-session <…>]` too.

AliSoftware committed:
Support `~/.aws/config` file parsing

Especially so that even if `region` is not defined in `~/.aws/credentials` but is defined in `~/.aws/config`, we can still create `AWSCredentials` with that `region` value.

AliSoftware committed:
Use Compile-Time RegEx Literals

Instead of Runtime String-evaluated ones that required a `try!`

AliSoftware committed:
Extract parsing from model + replace Bool w/ enum

AliSoftware committed:
Refactor fixture for easier-to-read tests

And to make Swiftlint happy with shorter line lengths

jkmassel commented on Sources/tinys3/model/AWSProfileConfig.swift:
```diff
@@ -0,0 +1,108 @@
+    ///   - path: The URL of the file to parse
+    ///   - isCredentialsFile: Indicates if the file to parse is a credentials file (true) or a config file (false)
+    /// - Returns: A dictionary containing an `AWSProfileConfig` for each profile name found
+    static func profiles(from path: URL, isCredentialsFile: Bool) throws -> [String: AWSProfileConfig] {
```
IMHO we should change `isCredentialsFile` to an enum like `fileType` to handle other potential cases (or even if it's only ever the two cases, it's easier to understand what's going on when branching on the file type.

jkmassel replied:
Addressed by afca280eeb5ecfadbd0f82ed0ee7de12cd11ee9b

jkmassel commented on Sources/tinys3/model/AWSProfileConfig.swift:
```diff
@@ -0,0 +1,108 @@
+    ///   - string: The content of the file to parse
+    ///   - isCredentialsFile: Indicates if the file to parse is a credentials file (true) or a config file (false)
+    /// - Returns: A dictionary containing an `AWSProfileConfig` for each profile name found
+    static func profiles(from string: String, isCredentialsFile: Bool) throws -> [String: AWSProfileConfig] {
```
Nit, but there's lots going on in this method – IMHO having a parser object is a bit nicer for testing in isolation and keeping things simple.

AliSoftware replied:
I think afca280 addresses this? 🙂

AliSoftware replied:
This was already extracted in a dedicated `struct AWSProfileConfigFileParser` type. To make this separation clearer, I've now since moved that type into its dedicated file in ec82c84f1d92c221123b8c17a3d5af6cb88cdaa2.

jkmassel commented on Tests/tinys3Tests/model/AWSCredentialsTests.swift:
```diff
@@ -3,13 +3,80 @@ import XCTest
-    let credentials: AWSCredentials = .testDefault
+    func testThatSingleFileContainsValidCredentials() throws {
+        let profiles = try AWSProfileConfig.profiles(from: R.AWSCredentialsFile.single, isCredentialsFile: true)
+        let defaultConfig = try XCTUnwrap(profiles["default"])
```
Nit – `default` is super common – can we have a type here to accept that?

AliSoftware replied:
Addressed in ec82c84f1d92c221123b8c17a3d5af6cb88cdaa2

AliSoftware committed:
Use dedicated type for AWSProfile names

 - `AWSProfile` wraps a profile name String in a type-safe way, which not only makes method signatures clearer in intent but also allows to use `.default` for the common `AWSProfile(name: "default")` case.
 - Also moves `AWSProfileConfigFileParser` to a dedicated file

AliSoftware committed:
Make swiftlint happy

iangmaia commented on Sources/tinys3/model/AWSProfileConfigFileParser.swift:
```diff
@@ -0,0 +1,99 @@
+    ///
+    /// - Parameters:
+    ///   - string: The content of the file to parse
+    ///   - isCredentialsFile: Indicates if the file to parse is a credentials file (true) or a config file (false)
```
```suggestion
    ///   - fileType: Indicates if the file to parse is a credentials file (.credentials) or a config file (.config)
```

AliSoftware replied:
Nice catch, thanks! Fixed in [018c403](https://github.com/Automattic/hostmgr/pull/93/commits/018c403ad26c6d72ce8a97ddbe03bbddf34dc3de)

iangmaia commented on Sources/tinys3/model/AWSProfileConfigFileParser.swift:
```diff
@@ -0,0 +1,99 @@
+    /// Parses the profiles and their associated configuration values from a given file
+    /// - Parameters:
+    ///   - path: The URL of the file to parse
+    ///   - isCredentialsFile: Indicates if the file to parse is a credentials file (true) or a config file (false)
```
```suggestion
    ///   - fileType: Indicates if the file to parse is a credentials file (.credentials) or a config file (.config)
```

AliSoftware committed:
Fix Doc-Comments

Co-authored-by: Ian Guedes Maia <iangmaia@users.noreply.github.com>

AliSoftware commented:
ping @jkmassel for another review 🙂  (especially to lift the "change requested" blocker if you agree that I addressed all your latest feedback)
